crates/core/tedge/src/cli/connect/c8y_direct_connection.rs

```diff
@@ -73,6 +73,13 @@ pub fn create_device_with_direct_connection(
                     break;
                 }
             }
+            Ok(Event::Incoming(Incoming::PubComp(rumqttc::PubComp { pkid: 2 }))) => {
+                publish_device_create_message(
+                    &mut client,
+                    &bridge_config.remote_clientid.clone(),
+                    device_type,
+                )?;
+            }
```

Comment:
So this is "to speed up the connection (sending a device creation message as soon as the first one is acknowledged, previously we waited for a heartbeat message)".

Okay. But why as reaction to message numbered `2` ?
- Is this the message sent on `Ok(Event::Incoming(Packet::SubAck(_)))`
- or sent on `Ok(Event::Outgoing(Outgoing::PingReq))`.

This is also relies on the fact that `publish_device_create_message` send the message with `QoS::ExactlyOnce` (this is actually the case).

So this is okay but a bit fragile. So I would consider:
- to add some comments
- to react independently of the pkid (as we only send creation message on this connection)
- to react also on `PubAck`.

I would even consider to send the message with QoS AtLeastOnce as we are now blindly re-sending the message if c8y is not fast enough.

Reply:
It's using `pkid: 2` as that is the first `PubComp` we receive (corresponding to the message we sent after receiving `SubAck`), and I was trying to avoid this code triggering itself as we should need to send exactly two messages to get Cumulocity to reply to say the device already exists.

Reply:
> I was trying to avoid this code triggering itself as we should need to send exactly two messages to get Cumulocity to reply to say the device already exists.

You can discuss this with @rina23q as she already tried different approaches for that purpose.

You can also discuss this with @reubenmiller as he might shed a new light to this issue.